File: src/python_gui/vdtrace_gui/trace_profile.py
from __future__ import annotations

from dataclasses import dataclass, replace

from .depth_filters import build_depth_filter_spec, ui_depth_to_runtime_text
from .models import TraceConfig


@dataclass(slots=True)
class TraceProfile:
    agent_path: str
    thread_id: str = "0"
    thread_capture: bool = True
    modules: str = "UnityPlayer.dll"
    output_path: str = ""
    max_events: str = "0"
    backend: str = "DR"
    call_depth: str = "3"
    outside_call_depth_enabled: bool = False
    outside_call_depth: str = "3"
    outside_execution_mode: str = "EDGE"
    anonymous_exec_call_depth_enabled: bool = False
    anonymous_exec_call_depth: str = "3"
    anonymous_exec_execution_mode: str = "EDGE"
    module_call_depths: str = ""
    trigger_point: str = "GameAssembly.dll+0x3498AE0"
    probe_spec: str = ""
    probe_enabled: bool = False
    trigger_enabled: bool = True
    block_main_thread: bool = False
    trace_outside_modules: bool = False
    all_events: bool = False
    repeat_hits: bool = False
    idle_escape_enabled: bool = True
    idle_escape_threshold: str = "32"
    enhanced_sampling: bool = False
    root_stop_on_return: bool = True
    async_thread_handoff: bool = True


def default_trace_profile(agent_path: str) -> TraceProfile:
    return TraceProfile(agent_path=agent_path)


def copy_trace_profile(profile: TraceProfile) -> TraceProfile:
    return replace(profile)


def normalize_backend_text(text: str) -> str:
    backend = text.strip().upper() or "DR"
    if backend in ("DR", "TF", "PT"):
        return backend
    return ""


def validate_probe_spec_text(text: str) -> str:
    stripped = text.strip()
    if not stripped:
        raise ValueError("已启用观测器，但规则为空。")

    for rule in [item.strip() for item in stripped.split(";") if item.strip()]:
        lowered = rule.lower()
        if lowered.startswith("step@") or lowered.startswith("write@"):
            tokens = rule.split()
            if not tokens or "@" not in tokens[0]:
                raise ValueError("观测器规则语法无效。")
            is_write = lowered.startswith("write@")
            has_steps = False
            has_watch = False
            for token in tokens[1:]:
                if "=" not in token:
                    raise ValueError("观测器规则语法无效：参数必须使用 key=value。")
                key, value = [item.strip() for item in token.split("=", 1)]
                key = key.lower()
                if key == "steps":
                    try:
                        parsed = int(value, 0)
                    except ValueError as exc:
                        raise ValueError("观测器规则语法无效：steps 不是正整数。") from exc
                    if parsed <= 0:
                        raise ValueError("观测器规则语法无效：steps 不是正整数。")
                    has_steps = True
                elif key == "exit":
                    if value.lower() not in ("return", "leave", "return-or-leave", "return_or_leave"):
                        raise ValueError("观测器规则语法无效：exit 只支持 return / leave / return-or-leave。")
                elif key == "watch":
                    if not is_write:
                        raise ValueError("step 规则不能带 watch。")
                    watches = [item.strip() for item in value.split("|") if item.strip()]
                    if not watches or len(watches) > 4:
                        raise ValueError("write 规则需要 1 到 4 个 watch。")
                    for watch in watches:
                        parts = [item.strip() for item in watch.split(":")]
                        if len(parts) < 2 or len(parts) > 3:
                            raise ValueError("write watch 语法无效。")
                        try:
                            size = int(parts[1], 0)
                        except ValueError as exc:
                            raise ValueError("write watch 大小无效。") from exc
                        if size <= 0 or size > 32:
                            raise ValueError("write watch 大小必须在 1 到 32 之间。")
                    has_watch = True
                else:
                    raise ValueError(f"未知的观测器参数：{key}")
            if not has_steps:
                raise ValueError("step/write 规则必须显式配置 steps。")
            if is_write and not has_watch:
                raise ValueError("write 规则必须显式配置 watch。")
            continue

        if "->" not in rule:
            raise ValueError("观测器规则语法无效：capture 规则缺少 '->'。")
        hit_text, capture_text = [item.strip() for item in rule.split("->", 1)]
        if not hit_text or not capture_text:
            raise ValueError("观测器规则语法无效：命中点或 capture 为空。")
        captures = [item.strip() for item in capture_text.split("|") if item.strip()]
        if not captures:
            raise ValueError("观测器规则语法无效：没有 capture。")
        if len(captures) > 4:
            raise ValueError("单个 capture 规则最多只能配置 4 个 capture。")
        for capture in captures:
            parts = [item.strip() for item in capture.split(":")]
            if len(parts) < 2:
                raise ValueError("capture 语法无效。")
            kind = parts[0].lower()
            if kind == "reg":
                if len(parts) > 3:
                    raise ValueError("reg capture 语法无效。")
            elif kind == "mem":
                if len(parts) < 3 or len(parts) > 4:
                    raise ValueError("mem capture 语法无效。")
            elif kind == "ptr":
                if len(parts) < 3 or len(parts) > 4:
                    raise ValueError("ptr capture 语法无效。")
            else:
                raise ValueError(f"未知的 capture 类型：{parts[0]}")
    return stripped


def build_trace_config_from_profile(profile: TraceProfile) -> TraceConfig:
    backend = normalize_backend_text(profile.backend) or "DR"
    if backend == "PT":
        raise ValueError("PT 已迁到 backup/pt；当前主线只支持 DR / TF。")
    trigger_enabled = profile.trigger_enabled
    thread_capture_enabled = profile.thread_capture
    thread_id = _parse_non_negative_int(profile.thread_id, "线程 ID")
    max_events = _parse_non_negative_int(profile.max_events, "事件上限")
    idle_escape_threshold = _parse_non_negative_int(profile.idle_escape_threshold, "空转跳出阈值")
    auto_select_thread = trigger_enabled and thread_capture_enabled
    block_main_thread = auto_select_thread and profile.block_main_thread
    if trigger_enabled:
        if auto_select_thread:
            thread_id = 0
    elif thread_capture_enabled:
        thread_id = 0
    probe_spec = validate_probe_spec_text(profile.probe_spec) if profile.probe_enabled else ""
    depth_filter_spec = build_depth_filter_spec(
        profile.outside_call_depth_enabled,
        profile.outside_call_depth,
        profile.outside_execution_mode,
        profile.anonymous_exec_call_depth_enabled,
        profile.anonymous_exec_call_depth,
        profile.anonymous_exec_execution_mode,
        profile.module_call_depths,
    )

    return TraceConfig(
        thread_id=thread_id,
        auto_select_thread=auto_select_thread,
        block_main_thread=block_main_thread,
        modules=profile.modules.strip(),
        output_path=profile.output_path.strip(),
        max_events=max_events,
        trace_outside_modules=profile.trace_outside_modules,
        backend=backend,
        control_flow_only=backend != "TF",
        max_call_depth=ui_depth_to_runtime_text(profile.call_depth),
        depth_filter_spec=depth_filter_spec,
        hit_policy="every" if profile.repeat_hits else "first",
        hot_bypass_threshold=idle_escape_threshold if profile.idle_escape_enabled else 0,
        enhanced_sampling=profile.enhanced_sampling,
        trigger_point=profile.trigger_point.strip() if trigger_enabled else "",
        probe_spec=probe_spec,
        stop_on_root_return=profile.root_stop_on_return,
        async_thread_handoff=profile.async_thread_handoff,
    )


def format_trace_profile(profile: TraceProfile) -> str:
    depth_filter_spec = build_depth_filter_spec(
        profile.outside_call_depth_enabled,
        profile.outside_call_depth,
        profile.outside_execution_mode,
        profile.anonymous_exec_call_depth_enabled,
        profile.anonymous_exec_call_depth,
        profile.anonymous_exec_execution_mode,
        profile.module_call_depths,
    )
    parts = [
        f"agent_path={profile.agent_path}",
        f"thread_id={profile.thread_id.strip() or '0'}",
        f"thread_capture={'true' if profile.thread_capture else 'false'}",
        f"modules={profile.modules.strip() or '-'}",
        f"output_path={profile.output_path.strip() or '(auto)'}",
        f"max_events={profile.max_events.strip() or '0'}",
        f"backend={normalize_backend_text(profile.backend) or 'DR'}",
        f"call_depth={profile.call_depth.strip() or '0'}",
        f"outside_execution_mode={profile.outside_execution_mode.strip() or 'EDGE'}",
        f"anonymous_exec_execution_mode={profile.anonymous_exec_execution_mode.strip() or 'EDGE'}",
        f"trigger_enabled={'true' if profile.trigger_enabled else 'false'}",
        f"trigger_point={profile.trigger_point.strip() or '-'}",
        f"observer_enabled={'true' if profile.probe_enabled else 'false'}",
        f"observer_spec={profile.probe_spec.strip() or '-'}",
        f"trace_outside_modules={'true' if profile.trace_outside_modules else 'false'}",
        f"repeat_hits={'true' if profile.repeat_hits else 'false'}",
        f"idle_escape_enabled={'true' if profile.idle_escape_enabled else 'false'}",
        f"idle_escape_threshold={profile.idle_escape_threshold.strip() or '32'}",
        f"enhanced_sampling={'true' if profile.enhanced_sampling else 'false'}",
        f"block_main_thread={'true' if profile.block_main_thread else 'false'}",
        f"root_stop_on_return={'true' if profile.root_stop_on_return else 'false'}",
        f"async_thread_handoff={'true' if profile.async_thread_handoff else 'false'}",
        f"depth_filter_spec={depth_filter_spec or '-'}",
    ]
    return "\n".join(parts)


def _parse_non_negative_int(text: str, field_name: str) -> int:
    stripped = text.strip() or "0"
    try:
        value = int(stripped)
    except ValueError as exc:
        raise ValueError(f"{field_name}必须是非负整数。") from exc
    if value < 0:
        raise ValueError(f"{field_name}必须是非负整数。")
    return value

File: src/flutter_gui/lib/src/trace_profile.dart
import 'depth_filters.dart';
import 'models.dart';

String normalizeBackendText(String text)
{
  final backend = text.trim().toUpperCase();
  if (backend.isEmpty)
  {
    return 'DR';
  }
  if (backend == 'DR' || backend == 'TF' || backend == 'PT')
  {
    return backend;
  }
  return '';
}

String validateProbeSpecText(String text)
{
  final stripped = text.trim();
  if (stripped.isEmpty)
  {
    throw FormatException('已启用观测器，但规则为空。');
  }
  for (final raw in stripped.split(';'))
  {
    final rule = raw.trim();
    if (rule.isEmpty)
    {
      continue;
    }
    final lowered = rule.toLowerCase();
    if (lowered.startsWith('step@') || lowered.startsWith('write@'))
    {
      _validateLocalProbeRule(rule, lowered.startsWith('write@'));
      continue;
    }
    _validateCaptureProbeRule(rule);
  }
  return stripped;
}

TraceConfig buildTraceConfigFromProfile(TraceProfile profile)
{
  final backend = normalizeBackendText(profile.backend).isEmpty ? 'DR' : normalizeBackendText(profile.backend);
  if (backend == 'PT')
  {
    throw FormatException('PT 已迁到 backup/pt；当前主线只支持 DR / TF。');
  }
  final triggerEnabled = profile.triggerEnabled;
  final threadCaptureEnabled = profile.threadCapture;
  var threadId = _parseNonNegativeInt(profile.threadId, '线程 ID');
  final maxEvents = _parseNonNegativeInt(profile.maxEvents, '事件上限');
  final idleEscapeThreshold = _parseNonNegativeInt(profile.idleEscapeThreshold, '空转跳出阈值');
  final autoSelectThread = triggerEnabled && threadCaptureEnabled;
  final blockMainThread = autoSelectThread && profile.blockMainThread;
  if (triggerEnabled)
  {
    if (autoSelectThread)
    {
      threadId = 0;
    }
  }
  else if (threadCaptureEnabled)
  {
    threadId = 0;
  }
  final probeSpec = profile.probeEnabled ? validateProbeSpecText(profile.probeSpec) : '';
  final depthFilterSpec = buildDepthFilterSpec(
    profile.outsideCallDepthEnabled,
    profile.outsideCallDepth,
    profile.outsideExecutionMode,
    profile.anonymousExecCallDepthEnabled,
    profile.anonymousExecCallDepth,
    profile.anonymousExecExecutionMode,
    profile.moduleCallDepths,
  );
  return TraceConfig(
    threadId: threadId,
    autoSelectThread: autoSelectThread,
    blockMainThread: blockMainThread,
    modules: profile.modules.trim(),
    outputPath: profile.outputPath.trim(),
    maxEvents: maxEvents,
    traceOutsideModules: profile.traceOutsideModules,
    backend: backend,
    controlFlowOnly: backend != 'TF',
    maxCallDepth: uiDepthToRuntimeText(profile.callDepth),
    depthFilterSpec: depthFilterSpec,
    hitPolicy: profile.repeatHits ? 'every' : 'first',
    hotBypassThreshold: profile.idleEscapeEnabled ? idleEscapeThreshold : 0,
    enhancedSampling: profile.enhancedSampling,
    triggerPoint: triggerEnabled ? profile.triggerPoint.trim() : '',
    probeSpec: probeSpec,
    stopOnRootReturn: profile.rootStopOnReturn,
    asyncThreadHandoff: profile.asyncThreadHandoff,
  );
}

String formatTraceProfile(TraceProfile profile)
{
  final depthFilterSpec = buildDepthFilterSpec(
    profile.outsideCallDepthEnabled,
    profile.outsideCallDepth,
    profile.outsideExecutionMode,
    profile.anonymousExecCallDepthEnabled,
    profile.anonymousExecCallDepth,
    profile.anonymousExecExecutionMode,
    profile.moduleCallDepths,
  );
  return <String>[
    'agent_path=${profile.agentPath}',
    'thread_id=${profile.threadId.trim().isEmpty ? '0' : profile.threadId.trim()}',
    'thread_capture=${profile.threadCapture}',
    'modules=${profile.modules.trim().isEmpty ? '-' : profile.modules.trim()}',
    'output_path=${profile.outputPath.trim().isEmpty ? '(auto)' : profile.outputPath.trim()}',
    'max_events=${profile.maxEvents.trim().isEmpty ? '0' : profile.maxEvents.trim()}',
    'backend=${normalizeBackendText(profile.backend).isEmpty ? 'DR' : normalizeBackendText(profile.backend)}',
    'call_depth=${profile.callDepth.trim().isEmpty ? '0' : profile.callDepth.trim()}',
    'trigger_enabled=${profile.triggerEnabled}',
    'trigger_point=${profile.triggerPoint.trim().isEmpty ? '-' : profile.triggerPoint.trim()}',
    'observer_enabled=${profile.probeEnabled}',
    'observer_spec=${profile.probeSpec.trim().isEmpty ? '-' : profile.probeSpec.trim()}',
    'trace_outside_modules=${profile.traceOutsideModules}',
    'repeat_hits=${profile.repeatHits}',
    'idle_escape_enabled=${profile.idleEscapeEnabled}',
    'idle_escape_threshold=${profile.idleEscapeThreshold.trim().isEmpty ? '32' : profile.idleEscapeThreshold.trim()}',
    'enhanced_sampling=${profile.enhancedSampling}',
    'block_main_thread=${profile.blockMainThread}',
    'root_stop_on_return=${profile.rootStopOnReturn}',
    'async_thread_handoff=${profile.asyncThreadHandoff}',
    'depth_filter_spec=${depthFilterSpec.isEmpty ? '-' : depthFilterSpec}',
  ].join('\n');
}

int _parseNonNegativeInt(String text, String fieldName)
{
  final value = int.parse(text.trim().isEmpty ? '0' : text.trim());
  if (value < 0)
  {
    throw FormatException('$fieldName必须是非负整数。');
  }
  return value;
}

void _validateLocalProbeRule(String rule, bool isWrite)
{
  final tokens = rule.split(RegExp(r'\s+')).where((token) => token.isNotEmpty).toList();
  if (tokens.isEmpty || !tokens.first.contains('@'))
  {
    throw FormatException('观测器规则语法无效。');
  }
  var hasSteps = false;
  var hasWatch = false;
  for (final token in tokens.skip(1))
  {
    if (!token.contains('='))
    {
      throw FormatException('观测器规则语法无效：参数必须使用 key=value。');
    }
    final index = token.indexOf('=');
    final key = token.substring(0, index).trim().toLowerCase();
    final value = token.substring(index + 1).trim();
    if (key == 'steps')
    {
      final parsed = int.parse(value);
      if (parsed <= 0)
      {
        throw FormatException('观测器规则语法无效：steps 不是正整数。');
      }
      hasSteps = true;
    }
    else if (key == 'exit')
    {
      if (!const {'return', 'leave', 'return-or-leave', 'return_or_leave'}.contains(value.toLowerCase()))
      {
        throw FormatException('观测器规则语法无效：exit 只支持 return / leave / return-or-leave。');
      }
    }
    else if (key == 'watch')
    {
      if (!isWrite)
      {
        throw FormatException('step 规则不能带 watch。');
      }
      final watches = value.split('|').map((item) => item.trim()).where((item) => item.isNotEmpty).toList();
      if (watches.isEmpty || watches.length > 4)
      {
        throw FormatException('write 规则需要 1 到 4 个 watch。');
      }
      for (final watch in watches)
      {
        final parts = watch.split(':').map((part) => part.trim()).toList();
        if (parts.length < 2 || parts.length > 3)
        {
          throw FormatException('write watch 语法无效。');
        }
        final size = int.parse(parts[1]);
        if (size <= 0 || size > 32)
        {
          throw FormatException('write watch 大小必须在 1 到 32 之间。');
        }
      }
      hasWatch = true;
    }
    else
    {
      throw FormatException('未知的观测器参数：$key');
    }
  }
  if (!hasSteps)
  {
    throw FormatException('step/write 规则必须显式配置 steps。');
  }
  if (isWrite && !hasWatch)
  {
    throw FormatException('write 规则必须显式配置 watch。');
  }
}

void _validateCaptureProbeRule(String rule)
{
  if (!rule.contains('->'))
  {
    throw FormatException("观测器规则语法无效：capture 规则缺少 '->'。");
  }
  final index = rule.indexOf('->');
  final hitText = rule.substring(0, index).trim();
  final captureText = rule.substring(index + 2).trim();
  if (hitText.isEmpty || captureText.isEmpty)
  {
    throw FormatException('观测器规则语法无效：命中点或 capture 为空。');
  }
  final captures = captureText.split('|').map((item) => item.trim()).where((item) => item.isNotEmpty).toList();
  if (captures.isEmpty)
  {
    throw FormatException('观测器规则语法无效：没有 capture。');
  }
  if (captures.length > 4)
  {
    throw FormatException('单个 capture 规则最多只能配置 4 个 capture。');
  }
  for (final capture in captures)
  {
    final parts = capture.split(':').map((part) => part.trim()).toList();
    if (parts.length < 2)
    {
      throw FormatException('capture 语法无效。');
    }
    final kind = parts.first.toLowerCase();
    if (kind == 'reg')
    {
      if (parts.length > 3)
      {
        throw FormatException('reg capture 语法无效。');
      }
    }
    else if (kind == 'mem' || kind == 'ptr')
    {
      if (parts.length < 3 || parts.length > 4)
      {
        throw FormatException('$kind capture 语法无效。');
      }
    }
    else
    {
      throw FormatException('未知的 capture 类型：${parts.first}');
    }
  }
}
